New layout for generic task forms

mockoocy opened the pull request:
The layout of the form was fixed to be a 2-column grid. Now it's still the default behaviour, but it can be changed. For each field it is possible to specify `uiSchema['ui:options'].col` parameter, to a number between 1 and 12 (or auto), see [bootstrap grid system](https://getbootstrap.com/docs/4.0/layout/grid/)

This way some more sophisticated form layouts can be achieved, like using some full-width custom widgets in future.

Most importantly, it fixes some visual bugs that were present in the past, like errors breaking the UI layout.

The form fields also got changed, from having labels on the left of input fields, to having them on top of the inputs. This makes it clearer when using more than 2 columns.


**Before**
![image](https://github.com/user-attachments/assets/cc2bdfa7-c6e0-4d42-b31f-db4c825e02a5)


**After** 
![image](https://github.com/user-attachments/assets/26c2ef9c-89fc-4110-8898-8fb05ff1e371)

mockoocy committed:
New layout for generic task forms

The layout of the form was fixed to be a 2-column grid.
Now it's still the default behaviour, but it can be changed.
For each field it is possible to specify `uiSchema['ui:options'].col`
parameter, to a number between 1 and 12 (or auto), see [bootstrap grid
system](https://getbootstrap.com/docs/4.0/layout/grid/)

This way some more sophisticated form layouts can be achieved, like
using some full-width custom widgets in future.

Most importantly, it fixes some visual bugs that were present in the
past, like errors breaking the UI layout.

The form fields also got changed, from having labels on the left of
input fields, to having them on top of the inputs.
This makes it clearer when using more than 2 columns.

axelboc commented on ui/src/components/Tasks/GenericTaskForm.jsx:
```diff
@@ -205,48 +205,56 @@ class GenericTaskForm extends React.Component {
-        {errors}
-        {help}
+      <div className={`${gridClass} ${classNames} field-container`.trim()}>
+        <div className="field-title">
```
Would you be able to use CSS modules for all the class names?

mockoocy replied:
Done - moved all the custom `field-*` classes to a css module.
The `field-container` class got also replaced by a *specific selector* `form-group field:not(.field-object)` which describes the same objects.
`RJFS`-related selectors are still in the plain `.css` file, as it uses the 'normal', global classes.

axelboc reviewed (commented):
Awesome, with the two columns, it's indeed much clearer to have the labels above the fields.

I notice you use conditional rendering like `rawDescription && ()`, which is great, but in the screenshot, it seems that there are still gaps when fields don't have a description:

![image](https://github.com/user-attachments/assets/09dd69c8-34ee-4a0f-b6ae-de3317c9d223)

Would it be able to remove these gaps somehow? Otherwise, it almost looks like the label belongs with the field above it... :joy:

marcus-oscarsson commented:
Looks quite nice, a good start !

I think it would be really nice if we could remove the space for the components that are not rendered, for instance as @axelboc mentions the description.

It would be nice if we could do something for the input size. I know that its, at the moment, decided by the column layout but it looks a bit strange to have a input field that takes half the dialog for an exposure time that is 2 digits. Not sure if there is a way to specify the input size and then default to something reasonable.

It would also be very nice if we could have some spacing between each row so that its clear what layouts and inputs that go together and it would also be easier on the eye.

mockoocy commented:
Thanks @marcus-oscarsson, @axelboc for the feedback.
Yeah, the vertical gaps were certainly too large there. I was just bothered with situations where one field has a description and the one next to it has no description.

 In that case the input fields were misaligned. so I decided to make the length of the row containing the label and description fixed. But now I agree that it made things worse :)

I've removed some of the vertical gaps now, and it looks like this for 2-col form.
![image](https://github.com/user-attachments/assets/7a1c6c1f-22d8-4f1d-98d6-abaa621973b7)

The input fields are probably also too wide, as Marcus pointed out.
There is no way to change the width currently, other than changing the amount of cols (the `ui:options.col` property from `uiSchema`)
If that's an issue, it could be made so that the width could be specified using `uiSchema`, if unspecified it would just take width of a whole column. What would be your idea?

axelboc reviewed (approved):
Thanks! Yeah, I don't think having the fields not lined up across the two columns is a huge deal. You could put the label and description on the same line with automatic wrapping, which would at least line things up when both the label and description are quite short like in your screenshot... No strong feelings about this, though.

I would maybe just add a tiny bit of space between labels and fields, and between rows, like 2 and 4px respectively maybe (to be tested).

marcus-oscarsson commented:
Already looks better, I also agree that some more space between the rows would be more pleasant. 

Would it be possible to use a tool-tip for the description and possibly the error as well, one for each with a nice little icon. Like that we would always render the rows the same ?,

It's not extremely disturbing but its somewhat annoying when the rows don't lineup.

Something that would be really neat is if we could create groups with different col layouts for each group and perhaps with an optional title. Somehow one would for instance want Cell A, B C be on one row and Cell alpha, beta, gamma on another. Space group on its own row and a nice title saying Crystal or something like that. Its perhaps pushing it a bit far for this PR, but it would be super nice.

mockoocy commented:
I think the idea with the tooltip is good. Will add it.

As per @marcus-oscarsson remark; I see that `rjsf` has added something like this in a new release (v6 beta): https://rjsf-team.github.io/react-jsonschema-form/docs/usage/layout-grid 
We could try using it in the future I imagine.

mockoocy committed:
Adds tooltip for field description in generic form

This also introduces refactor. The `customFieldTemplate` has been moved
to be a separate functional component.
This should make the `GenericTaskForm` component a bit thinner, while
also making a progress in the `class->functional` rewrite of the
components.

It also introduces a slight change in the global css.
By default bootstrap provides a higher z-index value for modals than for
tooltips. It is problematic for this use-case of describing form fields
in a modal.
So I've increased z-index for tooltips that are rendered by rjsf.

There was also a slight change. There is no conditional rendering for
`errors` property in `CustomFieldTemplate` - it is already a React
element - it's always truthy.

mockoocy commented:
I decided to keep the description, in a tooltip. Much better! :)
Also I did a slight 'refactor' there. I moved the component that renders the fields to be a separate functional component.


![image](https://github.com/user-attachments/assets/7a3d8db2-963d-4796-a125-c1e0897ccb7e)

axelboc commented on ui/src/components/Tasks/CustomFieldTemplate.jsx:
```diff
@@ -0,0 +1,71 @@
+  return (
+    <div className={`${gridClass} ${classNames}`.trim()}>
+      <div className={styles.fieldTitle}>
+        {label ? (
```
I preferred the `&&` syntax that you had before :grin:

mockoocy replied:
Well, for some reason vscode automatically re-formats it using ternary operator :D

mockoocy replied:
I just checked; it conflicts with `react/jsx-no-leaked-render` eslint rule.
so I used the `!!var && (<Component/>)` syntax :)

axelboc replied:
Right, I actually ended up turning off that rule in the shared config: https://gitlab.esrf.fr/ui/eslint-config/-/merge_requests/4 — if you feel like upgrading, go for it, otherwise no worries.

mockoocy replied:
Oh, it just seems that I had outdated dependencies on my local environment. Now it works perfectly, thanks :)
Moved all these ternaries to use `&&` instead.

axelboc commented on ui/src/components/Tasks/style.css:
```diff
@@ -18,33 +18,14 @@
+  Select any element whose id starts with "root_" AND ends with "_tooltip" 
+  These should corresponds to tooltips generated for the GenericTaskForm
+  */
+[id^='root_'][id$='_tooltip'] {
```
Not a huge fan of this. I assume it's a bit tricky to pass a custom class name to `TooltipTrigger`. What about a `data` attribute like `data-tooltip-in-modal`?

mockoocy replied:
Oh, yeah - that's much better idea

axelboc reviewed (commented):
Even better! Kudos on the TypeDoc comment :clap:

marcus-oscarsson commented on ui/src/components/Tasks/CustomFieldTemplate.jsx:
```diff
@@ -0,0 +1,75 @@
+import styles from './CustomFieldTemplate.module.css';
+
+/**
+ * @typedef {Object} CustomFieldTemplateProps
```
Nice :)

marcus-oscarsson commented on ui/src/components/Tasks/style.css:
```diff
@@ -18,33 +18,11 @@
-
-.json-schema-form-group-div .field-description {
-  display: none;
+.tooltip[data-tooltip-in-modal='true'] {
```
yeah, thats easier to read :)

axelboc replied:
FYI, if you want to just write `.tooltip[data-tooltip-in-modal]`, you can do this in the JSX: `data-tooltip-in-modal={inModal || undefined}`

marcus-oscarsson commented:
Very nice, it would be really nice if we could look into the column layout in RJSF V6. Its probably better to do that in a separate PR though.

marcus-oscarsson commented:
The errors are still displayed inline I guess, do we want to keep it that way for the moment then ?

mockoocy commented:
> The errors are still displayed inline I guess, do we want to keep it that way for the moment then ?

I think for now it is good enough. Maybe in the future we could remove the list of errors on top of the form (unless it has some valid use case)  and perhaps think of a nicer way of displaying these.

axelboc commented:
Agreed that it's good enough for now and that the errors at the top could probably be removed later on. I think errors below the fields are fine (except perhaps for some minor styling tweaks later on), since they shouldn't ever be visible when the modal appears, so it doesn't matter too much if the layout shifts a bit once they start appearing.

jbflo commented:
> I decided to keep the description, in a tooltip. Much better! :) Also I did a slight 'refactor' there. I moved the component that renders the fields to be a separate functional component.
> 
> ![image](https://private-user-images.githubusercontent.com/67100892/448299272-7a3d8db2-963d-4796-a125-c1e0897ccb7e.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDg0MzI4MzQsIm5iZiI6MTc0ODQzMjUzNCwicGF0aCI6Ii82NzEwMDg5Mi80NDgyOTkyNzItN2EzZDhkYjItOTYzZC00Nzk2LWExMjUtYzFlMDg5N2NjYjdlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTI4VDExNDIxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZhZjM3ZDg0ODg1NzVmODg0OGQ3NjNlYzBkNDcwZmQ4YzYxNmY3YWQ1ZmQwZGY3ODRkOGE1NTQxNTJmMjY3MTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YSYMIEIpUn69AcjWM0JdKXX0kk3gcG36XZ9SvGhR070)

A little margin top on the labels maybe ?  the Labels are a bit too close to the fields on top.

mockoocy commented:
> A little margin top on the labels maybe ? the Labels are a bit too close to the fields on top.

Good idea. 
![image](https://github.com/user-attachments/assets/8011624b-9d38-470e-bbe8-2a0b0a893694)

marcus-oscarsson commented:
Starts to look pretty nice. I think its fine for this iteration. I do think we should think about the error display but thats for another discussion.  Another thing that would be really nice is to remove the generic task from from the redux-form infrastructure.

marcus-oscarsson commented:
There is long weekend in France starting from tomorrow. Feel free to merge if you are done ! Thanks again

mockoocy committed:
Makes error list prettier in generic task forms

Uses RJSF `FieldErrorTemplate` to create a custom component which
renders list of errors.
When an error occurs in an input field, then it is indicated by a red
border now.

Also reduced margin around the help message, to reduce visual clutter.

Additionally, changed JSDoc types, to use types from the library

mockoocy commented:
I've created an issue to move the `GenericTaskForm` to use `react-hook-form` instead of `redux-form`, https://github.com/mxcube/mxcubeweb/issues/1660.

Also, I included some additional changes, that you'll be able to review after the long weekend :)

First of all, I changed the JSDoc type hints, to use type definitions directly from the `rjsf` library. 
Also, I decided to include changes to how errors are displayed in this PR afterall. This is still somewhat relevant to this PR, and required me to change the css that I've introduced here.

![image](https://github.com/user-attachments/assets/e7b35187-1c38-4a99-8bf1-6e347db5a166)

marcus-oscarsson commented:
Looks really nice :)

marcus-oscarsson commented:
Thanks for creating the issue !

marcus-oscarsson commented on ui/src/components/Tasks/CustomFieldTemplate.jsx:
```diff
@@ -0,0 +1,61 @@
+/**
+ * CustomFieldTemplate is a custom field template for React JSON Schema Form (RJSF).
+ * It is used to render fields with a specific layout and styling.
+ * @param {import('@rjsf/utils').FieldTemplateProps} props
```
Oh, wow I see

axelboc commented on ui/src/components/Tasks/CustomFieldErrorTemplate.jsx:
```diff
@@ -0,0 +1,34 @@
+
+export default function CustomFieldErrorTemplate(props) {
+  const { errors = [], idSchema } = props;
+  // const fieldName = schema.title || schema.name || 'Field';
```
commented code

axelboc commented on ui/src/components/Tasks/CustomFieldErrorTemplate.jsx:
```diff
@@ -0,0 +1,34 @@
+      {errors.map((error, idx) => (
+        <div
+          // eslint-disable-next-line react/no-array-index-key
+          key={`${idSchema.$id}_error_${idx}`}
```
There's no benefit to generating complicated keys here. They just need to be unique within their parent element (and ideally not based on an array index). I would simply use the error message as key: `key={error}`, as it is likely unique. (and remove the `idSchema` prop altogether)

axelboc commented on ui/src/components/Tasks/CustomFieldErrorTemplate.jsx:
```diff
@@ -0,0 +1,34 @@
+          // eslint-disable-next-line react/no-array-index-key
+          key={`${idSchema.$id}_error_${idx}`}
+          className={`${styles.errorMessage} ${
+            idx < errors.length - 1 ? styles.errorMessageWithBorder : ''
```
I would typically use a `data` attribute for this, like `data-with-border={idx < errors.length - 1 || undefined}`, and then `.errorMessage[data-with-border]`.

That being said, I think it'd look better without the border to be honest.

axelboc commented on ui/src/components/Tasks/CustomFieldErrorTemplate.jsx:
```diff
@@ -0,0 +1,34 @@
+          }`}
+        >
+          <BsExclamationTriangle className={styles.errorIcon} />
+          <span>{error}</span>
```
Is the `span` needed?

mockoocy replied:
No, it was just a leftover.

axelboc commented on ui/src/components/Tasks/CustomFieldTemplate.jsx:
```diff
@@ -0,0 +1,61 @@
+            id={`${id}_tooltip`}
+            inModal
+          >
+            <span>
```
`span` needed?

mockoocy replied:
Yes, it's needed. 
The children to `OverlayTrigger` must accept a ref, so a plain `svg` tag is not sufficient. 
https://react-bootstrap.netlify.app/docs/components/overlays/#overlaytrigger

axelboc commented on ui/src/components/Tasks/CustomFieldErrorTemplate.jsx:
```diff
@@ -0,0 +1,34 @@
+ * @param {import('@rjsf/utils').FieldErrorProps} props
+ * @returns
+ */
+
```
Likely doesn't matter, but maybe remove that extra blank line?

axelboc reviewed (commented):
Even better :+1:

axelboc commented:
I'm re-approving obviously. :wink: Next time, don't hesitate to click on "re-request review" if you think the new changes deserve another look over so I know to prioritise a follow-up review:

![image](https://github.com/user-attachments/assets/0e6daafa-8bef-4c7d-b075-97adf8712c5c)

marcus-oscarsson commented:
@mockoocy I let you decide what to do with the last comments, you can merge when you feel that you are done. Unless you need to add some major changes of course.

mockoocy commented:
> @mockoocy I let you decide what to do with the last comments, you can merge when you feel that you are done. Unless you need to add some major changes of course.

Well, I don't have permissions to merge actually, so I will let @axelboc have one last look and merge it :)

marcus-oscarsson commented:
@mockoocy You should be able to merge after you accepted the invitation I just sent you ;)

mockoocy commented:
> @mockoocy You should be able to merge after you accepted the invitation I just sent you ;)

It works, thanks ;)
